Issue #14, #18, #19

cw646 opened the pull request:
Included maps for vertices and elements, and method to use vertices to generate points within an element using Eigen library.

cw646 committed:
:construction: Working on issue #18

cw646 committed:
:construction: Working on issue #14, #18, #19
Need further investigation into stress issues, stress currently set @ 0.

cw646 committed:
:construction: Working on issue #14, #18, #19
Removal of Eigen folder, include changes to cmake to support docker Eigen.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/cb-geo/mpm-point-generator?pullRequest=20) <br/>All committers have signed the CLA.

kks32 commented on include/gmsh.h:
```diff
@@ -5,12 +5,17 @@
 
+//libraries
+#include <Eigen/Dense>
+using Eigen::MatrixXd;
```
Avoid shortening of namespace, use `Eigen::MatrixXd` everywhere to avoid namespace clash.

kks32 commented on include/gmsh.h:
```diff
@@ -21,17 +26,26 @@ class GMSH : public Mesh {
   //! Read elements in GMSH
   void read_elements(const std::string& filename);
 
+  //!Store element id and vertices coordinates as map
```
You need a space after `//!`.

kks32 commented on include/gmsh.h:
```diff
@@ -40,15 +54,21 @@ class GMSH : public Mesh {
   std::vector<std::array<double, 6>> stress_;
+
+  //! Map to store id and vertices coordinates
+  std::map<double, std::array<double, 3>> vertices_;
```
Why is the `id` a double instead of an unsigned?

kks32 commented on include/gmsh.h:
```diff
@@ -40,15 +54,21 @@ class GMSH : public Mesh {
+
+  //! Map to store element ID and vertices ID
+  // 4 = element vertices
+  std::map<double, std::array<double, 4>> elements_;
```
ids need to be unsigned

kks32 commented on include/gmsh.tcc:
```diff
@@ -109,23 +111,143 @@ void GMSH::read_elements(const std::string& filename) {
-        if (elementtype != ndimension) {
+        //! \details If element type not == to specified elementtype_, skip
+        //! element
+        if (elementtype != elementype) {
```
Is this right? Should it be `elementtype != elementtype_`

cw646 replied:
Ah yes, I removed the trailing _ in previous commit, but should rename it to something else.

kks32 commented on include/gmsh.tcc:
```diff
@@ -109,23 +111,143 @@ void GMSH::read_elements(const std::string& filename) {
+      }
+      elementcoordinates_.insert(
+          std::make_pair(elementfind->first, verticesarray));
+    } else {
```
Why an empty else?

kks32 commented on include/gmsh.tcc:
```diff
@@ -109,23 +111,143 @@ void GMSH::read_elements(const std::string& filename) {
+            verticesarray[0] = verticesfind->second[0];
+            verticesarray[1] = verticesfind->second[1];
+            verticesarray[2] = verticesfind->second[2];
+          } else {
```
Remove empty elses.

kks32 commented on include/gmsh.tcc:
```diff
@@ -109,23 +111,143 @@ void GMSH::read_elements(const std::string& filename) {
+      elementkeyvalues[2] = elementfind->second[2];
+      elementkeyvalues[3] = elementfind->second[3];
+
+      for (unsigned q = 0; q < 4; ++q) {
```
What's the 4, 12, make them `const`s

kks32 commented on include/gmsh.tcc:
```diff
@@ -109,23 +111,143 @@ void GMSH::read_elements(const std::string& filename) {
+
+void GMSH::compute_material_points() {
+
+  double xcoord;
```
Can you not use a 3D array?

kks32 commented on include/gmsh.tcc:
```diff
@@ -109,23 +111,143 @@ void GMSH::read_elements(const std::string& filename) {
+  double zcoord;
+
+  std::array<double, 3> pointsarray;
+  std::map<double, std::array<double, 12>>::iterator coordinatesfind;
```
What's 12?

kks32 commented on include/gmsh.tcc:
```diff
@@ -109,23 +111,143 @@ void GMSH::read_elements(const std::string& filename) {
+
+      // Centroid test
+      //++++++++++++++++++++++++++++++++++++++
+      xcoord = 0.25 * (m(0, 0) + m(0, 1) + m(0, 2) + m(0, 3));
```
Write it directly to `pointsarray`.

kks32 replied:
Can you not put this inside a loop?

kks32 commented on include/gmsh.tcc:
```diff
@@ -109,23 +111,143 @@ void GMSH::read_elements(const std::string& filename) {
+  const unsigned firstelementcoord = elementcoordinates_.begin()->first;
+  const unsigned lastelementcoord = elementcoordinates_.rbegin()->first;
+
+  for (unsigned t = firstelementcoord; t < lastelementcoord + 1; ++t) {
```
We need to rethink this algorithm. Is this through all the element vertices?

kks32 commented on src/main.cc:
```diff
@@ -18,12 +18,14 @@ int main(int argc, char** argv) {
-    //! Read mesh and compute stresses
+    //! Read mesh and compute material points & stresses
     mesh->read_vertices(io->mesh_file_name());
+    mesh->read_elements(io->mesh_file_name());
```
Read elements should call `store_elements` at the end of the function, no reason to call another function in main.

ezrayst commented on include/io.tcc:
```diff
@@ -1,21 +1,21 @@
     //! Note that for 2D, z values are 0
     //! For 1D, both y and z values are 0
-    for (auto const& vertex : vertices) {
+    for (auto const& vertex : materialpoints) {
```
is it better to use `point` instead of `vertex` here?

ezrayst commented on include/io.tcc:
```diff
@@ -24,12 +24,13 @@ void IO::write_vertices(
 //! \brief Write initial stresses of material points
 //! \param[in[ stresses Initial stress of material points
 void IO::write_stresses(const std::vector<std::array<double, 6>>& stresses) {
+  double id = 1;
```
any specific reason you used `double` instead `unsigned`

ezrayst commented on include/io.tcc:
```diff
@@ -40,14 +41,17 @@ void IO::write_stresses(const std::vector<std::array<double, 6>>& stresses) {
+      stress_file << id << '\t';
       for (double stress_component : stress) {
-        stress_file << stress_component << '\t';
+        stress_file << 0 << '\t';
```
I will handle this part

ezrayst commented on include/gmsh.tcc:
```diff
@@ -66,9 +64,13 @@ void GMSH::read_elements(const std::string& filename) {
 
   const unsigned toplines = 4;
   const unsigned ndimension = 3;
+  // specify element type 4 = tetrahedral
```
just wondering. isn't it `quadrilateral?`

cw646 replied:
For unstructured 2D it's triangles, and for unstructured 3D it's 4 node-tetrahedral.

Commit:
:wrench: fix issue #20 for stress

Commit:
:art: add travis

Commit:
:art: improve issue #20, for store elements

cw646 committed:
:construction: Working on issue #14, #18, #19
Inclusion of loop for storing coordinates, solved stress' not working with MPM.
